### PR TITLE
Fix collections warnings

### PR DESCRIPTION
**Pull request by 1fish2**
Fixes the scary deprecation warnings that clients see, as mentioned in #6755, and some of the other warnings in the code.

* The `compat/__init__.py` DeprecationWarning about 'collections' was already fixed in 1b9804b65ecd008cfb424752e65aced92b63ff92, not yet on PyPI.
* The `misc/frozendict.py` DeprecationWarning about 'collections' just needed to add `'Callable'` to `__all__` in that module.

For a speedy way to test this fix, make a pyenv virtualenv, use this `requirements.txt` file (my best guess on it) to install pips, and run a limited nose test:

```bash
pip install --upgrade pip setuptools virtualenv virtualenvwrapper virtualenv-clone wheel
pip install -r requirements.txt && pyenv rehash
bin/theano-nose theano/tests/test_config.py
```

I ran the limited and the full `bin/theano-nose` in Python 2.7.18, 3.8.5, and 3.9.0b5. (I tried to run it in Python 3.10-dev but `pip install numpy` failed.) The full test can take 1.5 - 3 hours.

I'm unsure how far to go in this PR on fixing other warnings, but Python 3.9 is stricter about regex patterns so I fixed some problems there.

### CODE REVIEW

File: theano/gpuarray/tests/test_subtensor.py [hunk 1]
```diff
@@ -321,7 +321,7 @@ def test_tensor_float16(self):
 
 class TestGpuAllocDiag(test_basic.TestAllocDiag):
     def __init__(self, name):
-        return test_basic.TestAllocDiag.__init__(
+        test_basic.TestAllocDiag.__init__(
```
Comment (twiecki) [inline theano/gpuarray/tests/test_subtensor.py:324]: I guess this doesn't return anything?

**Comment by twiecki** [comment]
Thanks @1fish2! This looks good. For the `collections` import, could we just drop Python <= 3.6 and get rid of it?

File: bin/theano_nose.py [hunk 2]
```diff
@@ -204,6 +205,8 @@ def main():
     if '--help' in sys.argv or '-h' in sys.argv:
         help()
     else:
+        warnings.simplefilter("default")  # Enable warnings before importing theano
+        os.environ["PYTHONWARNINGS"] = "default"  # Also affect subprocesses
```
Comment (nouiz) [inline bin/theano_nose.py:209]: I think this should not overwrite that variable if it is already defined by the user.

Comment (twiecki) [reply bin/theano_nose.py:209]: Fixed in https://github.com/Theano/Theano/pull/6760.

File: theano/gpuarray/tests/test_subtensor.py (hunk 1 above)
Comment (1fish2) [reply theano/gpuarray/tests/test_subtensor.py:324]: Conceptually, an `__init__()` method is not allowed to return anything, which technically means it's required to return `None`, which is what happens by default without an explicit `return` value. https://docs.python.org/3/reference/datamodel.html#object.__init__

So this change clarifies the situation and removes a PyCharm inspection warning.

**Comment by 1fish2** [comment]
Thank you, @twiecki !

> For the `collections` import, could we just drop Python <= 3.6 and get rid of it?

TL;DR: Can't drop Python 2.7.

It's OK to drop Python 3.0 to 3.3. Some of those earlier versions of Python 3 are hard to even install now. The [PyPI download stats for Theano](https://pypistats.org/packages/theano) shows -- for the last week of data, July 13-19 -- no downloads for Python 3.3 and it doesn't even track 3.0 - 3.2.

But Python 2.7 accounts for about 1/3 of Theano downloads last week. It turns out to be a really big job to move a project from Python 2.7 to 3, so a lot of organizations are still not there. I just spent 2+ months doing it for one project, and that was after a lot of background steps.

The key point: The only practical way to move an application to Python 3 is to incrementally change things until all its parts run in both Python 2.7 and (say) 3.8 then finish testing & debugging. Then they can drop Python 2.7. The long transition requires Python libraries that work in both Python 2.7 and 3.8 and ideally produce identical results.

Even at that, the team has to change the behavior on Python 2.7 to match changes forced by Python 3. To take one example, Python 3 adds "salt" to every string hash as a security measure, but that makes the values differ on every run and always differ from Python 2.7, so it's no longer useful for a pseudo-random number seed.

**Comment by twiecki** [comment]
Thanks for the fixes and explanations @1fish2. Theano 1.0.5 is now released.

File: bin/theano_nose.py (hunk 2 above)
Comment (1fish2) [reply bin/theano_nose.py:209]: This is a good idea. Perhaps the ideal would be to bump `simplefilter()` and `PYTHONWARNINGS` up to `default` if they're weaker, thus preserving stronger filters, but there doesn't seem to be an API to find out what the warnings filter contains.

So it could be worth moving the `warnings.simplefilter("default")` call into the same `if` statement, that is, set both only if the environment variable was unset.

**Comment by nouiz** [comment]
About Python 2.7, if you release a Theano 1.1, it would be fine to drop it. Most other lib dropped it already. They can still use previous Theano version on 2.7. They just won't get the updates. Just my personal opinion. If NumPy drop it, then taking the few times spend on Theano should be spend elsewhere then supporting Python 2.7.
